### ISSUE TITLE
Problem with CORS headers

### ISSUE THREAD

**Issue by operutka** [miguelgrinberg/python-engineio#52]
Hi,
I have a problem in engineio 1.5.4 with CORS headers not being sent when upgrading a connection to web sockets. It causes problems in web browsers (namely Firefox) as they reject the connection upgrade.

I have resolved the issue by disabling connection upgrades but I'd like to enable it again when it's fixed.

**Comment by miguelgrinberg** [comment]
What async mode are you using?

**Comment by operutka** [comment]
I use aiohttp. However, I'm afraid there might be still some cases when the CORS headers are not sent even though I've disabled connection upgrades. The browser starts complaining about CORS when the session is active for a longer period of time.

Note that I don't use engineio directly. I use socketio instead and every time when event processing takes long period of time (minutes), I see this problem.

I've managed to capture the following responses with missing CORS headers using Wireshark:

```
HTTP/1.1 500 Internal Server Error
Content-Type: text/html; charset=utf-8
Content-Length: 141
Date: Fri, 09 Jun 2017 09:34:00 GMT
Server: Python/3.5 aiohttp/2.1.0
Connection: close
```

The HTTP 500 response was sent 60 seconds after receiving a GET request.

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: upgrade
Sec-Websocket-Accept: 2U52pq/YCOSFjYwiodJ9E8LoDHo=
Transfer-Encoding: chunked
Content-Type: application/octet-stream
Date: Fri, 09 Jun 2017 09:51:37 GMT
Server: Python/3.5 aiohttp/2.1.0
```

The HTTP 101 was received when upgrading the protocol to web socket.

**Comment by operutka** [comment]
BTW I also see the following unhandled exception when the problem happens:

```
Traceback (most recent call last):
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_socket.py", line 19, in poll
    self.server.ping_timeout)]
  File "/usr/lib/python3.5/asyncio/tasks.py", line 396, in wait_for
    raise futures.TimeoutError()
concurrent.futures._base.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_server.py", line 137, in handle_request
    packets = await socket.handle_get_request(environ)
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_socket.py", line 78, in handle_get_request
    raise e
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_socket.py", line 75, in handle_get_request
    packets = await self.poll()
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_socket.py", line 22, in poll
    raise exceptions.QueueEmpty()
engineio.exceptions.QueueEmpty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/xxx/.local/lib/python3.5/site-packages/aiohttp/web_protocol.py", line 421, in start
    resp = yield from self._request_handler(request)
  File "/home/xxx/.local/lib/python3.5/site-packages/aiohttp/web.py", line 303, in _handle
    resp = yield from handler(request)
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_server.py", line 144, in handle_request
    await self.disconnect(sid)
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/asyncio_server.py", line 94, in disconnect
    await self._get_socket(sid).close()
  File "/home/xxx/.local/lib/python3.5/site-packages/engineio/server.py", line 411, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'
```

**Comment by miguelgrinberg** [comment]
@operutka what version of Firefox are you using? I'm testing 54.0 here, and it appears it does not enforce the same origin policy on WebSocket requests. Chrome and Safari also allow different origins on WebSocket.

I actually have no problem connecting a client server from one origin, to a aiohttp Socket.IO server running on a different one. Please give me more specific details about the test that you did and how I can reproduce it here.

**Comment by operutka** [comment]
@miguelgrinberg sorry for not responding, I was too busy. I'm also using FF 54.0 here.

However, I'm not sure if the problem is in the WebSocket requests. The disconnections might be caused by the HTTP 500 responses. I just know that the browser starts complaining about CORS headers not being sent and that the client disconnects and reconnects again. The problem is that the client gets a new session ID on every reconnection, so I cannot send it any response because I cannot pair it with the original session ID.

It's quite easily reproducible. Create a Socket.IO server using python-socketio. The server should wait for more than 60 seconds (e.g. 90) for every connected client and then send back a message. You'll see the same problems as I did when you connect to this server using a JS Socket.IO client running in FF.

Unfortunately, I cannot help you with testing anymore as I was forced to switch to plain WebSockets.

**Comment by miguelgrinberg** [comment]
> The server should wait for more than 60 seconds

How do you wait? That's important, you could be blocking other asynchronous tasks if you do a busy wait.

**Comment by operutka** [comment]
I'm not that stupid. :) You can do either:

```
await asyncio.sleep(...)
await send_response(sid)
```

or even:

```
async def delayed_send():
    await asyncio.sleep(...)
    await send_response(sid)

event_loop.create_task(delayed_send())
```

in your event handler. Both will give you the same result.

**Comment by miguelgrinberg** [comment]
Well, this might be all water under the bridge if you are unable to run more tests, but what you are doing is actually not a great idea. The event handlers for a given client are serialized, to avoid potential race conditions or other side effects if two handlers for the same user are running in parallel. So sleeping for 90 seconds in an event handler could actually block other events for that client.

The eventlet and gevent async modes have a mode in which you can turn off this protection and allow handlers to run in parallel. I have it in my task list to implement the same thing for asyncio.

**Comment by operutka** [comment]
I believe that this is not the case. First, my client won't send any more messages until it receives a response, so there's never more than one event for one client at the same time. Second, and as I've already mentioned, the problem won't disappear even if you do it like this:
```
async def delayed_send():
    await asyncio.sleep(...)
    await send_response(sid)

event_loop.create_task(delayed_send())
```

Note that this code does not block the event handler. It creates a separate task instead.

Of course, in real world, my server does not do any sleep. :) It has a lot of asynchronous work to do before it's able to send a response to the client. This is just a model example that should help you to reproduce the problem.

**Comment by miguelgrinberg** [comment]
@operutka the fix for the "session is disconnected" error you referenced above went in today in release 1.7.0. I'm trying to figure out if there are any remaining issues here. It is possible that I missed other cases where HTTP responses go out without the CORS headers. I think this would happen for any responses that I do not handle directly, so for example, cases where an error occurs and it bubbles up all the way to the framework, and it is the framework issuing the response. I think for non-error use cases all the responses have the configured CORS headers, but if you still have reproducible cases of this problem after you upgrade please let me know.

**Comment by operutka** [comment]
Thanks @miguelgrinberg! I'm sorry for my late response. I'll take a look at it, however, I'm not sure when exactly. Our server uses a different solution now. I suggest to close this issue and I'll open a new one if I find any problem.

**Comment by miguelgrinberg** [comment]
Sounds good. Thanks.
